`assert_{frame,series}_equal` incorrectly says structs with outer nulls are not equal

### Checks

- [x] I have checked that this issue has not already been reported.
- [x] I have confirmed this bug exists on the [latest version](https://pypi.org/project/polars/) of Polars.

### Reproducible example

```python
import polars as pl
from polars.testing import assert_frame_equal

lf = pl.LazyFrame({"a": [{"x": 0}]})

q1 = lf.select(pl.col.a.slice(1, 1).first())
q2 = lf.group_by(pl.lit(1)).agg(pl.col.a.slice(1, 1).first()).drop("literal")
assert_frame_equal(q1, q2)
```


### Log output

```shell
AssertionError: DataFrames are different (value mismatch for column "a")
[left]: shape: (1,)
Series: 'a' [struct[1]]
[
	null
]
[right]: shape: (1,)
Series: 'a' [struct[1]]
[
	null
]
```

### Issue description

These are equal, but the assertion fails.

### Expected behavior

Assertion does not trigger.

### Installed versions

<details>

```
1.35.2
```

</details>

--- a/crates/polars-core/src/chunked_array/comparison/mod.rs
+++ b/crates/polars-core/src/chunked_array/comparison/mod.rs
@@ -3,7 +3,7 @@ mod scalar;
 #[cfg(feature = "dtype-categorical")]
 mod categorical;
 
-use std::ops::{BitAnd, Not};
+use std::ops::{BitAnd, BitOr, Not};
 
 use arrow::array::BooleanArray;
 use arrow::bitmap::{Bitmap, BitmapBuilder};
@@ -812,30 +812,28 @@ where
         if is_missing && (a.has_nulls() || b.has_nulls()) {
             // Do some allocations so that we can use the Series dispatch, it otherwise
             // gets complicated dealing with combinations of ==, != and broadcasting.
-            let default = || {
-                BooleanChunked::with_chunk(PlSmallStr::EMPTY, BooleanArray::from_slice([true]))
-                    .into_series()
-            };
-            let validity_to_series = |x| unsafe {
+            let default =
+                || BooleanChunked::with_chunk(PlSmallStr::EMPTY, BooleanArray::from_slice([true]));
+            let validity_to_ca = |x| unsafe {
                 BooleanChunked::with_chunk(
                     PlSmallStr::EMPTY,
                     BooleanArray::from_inner_unchecked(ArrowDataType::Boolean, x, None),
                 )
-                .into_series()
             };
 
-            out = reduce(
-                out,
-                op(
-                    &a.rechunk_validity()
-                        .map_or_else(default, validity_to_series),
-                    &b.rechunk_validity()
-                        .map_or_else(default, validity_to_series),
-                ),
-            )
+            let a_s = a.rechunk_validity().map_or_else(default, validity_to_ca);
+            let b_s = b.rechunk_validity().map_or_else(default, validity_to_ca);
+
+            let shared_validity = (&a_s).bitand(&b_s);
+            let valid_nested = if op_is_ne {
+                (shared_validity).bitand(out)
+            } else {
+                (!shared_validity).bitor(out)
+            };
+            out = reduce(op(&a_s.into_series(), &b_s.into_series()), valid_nested);
         }
 
-        if !is_missing && (a.null_count() > 0 || b.null_count() > 0) {
+        if !is_missing && (a.has_nulls() || b.has_nulls()) {
             let mut a = a;
             let mut b = b;
 
@@ -894,7 +892,7 @@ impl ChunkCompareEq<&StructChunked> for StructChunked {
             self,
             rhs,
             |l, r| l.not_equal_missing(r).unwrap(),
-            |a, b| a | b,
+            |a, b| a.bitor(b),
             true,
             false,
         )
@@ -905,7 +903,7 @@ impl ChunkCompareEq<&StructChunked> for StructChunked {
             self,
             rhs,
             |l, r| l.not_equal_missing(r).unwrap(),
-            |a, b| a | b,
+            |a, b| a.bitor(b),
             true,
             true,
         )

--- a/py-polars/tests/unit/datatypes/test_struct.py
+++ b/py-polars/tests/unit/datatypes/test_struct.py
@@ -1389,3 +1389,29 @@ def test_struct_reverse_chunked_25269() -> None:
     assert_series_equal(
         pl.concat([s.head(1), s.tail(1)]).reverse(), pl.Series([{"a": None}, None])
     )
+
+
+def test_struct_equal_missing_null_25360() -> None:
+    lf = pl.LazyFrame({"a": [{"x": 0}]})
+
+    q1 = lf.select(a1=pl.col.a.slice(1, 1).first())
+    q2 = lf.group_by(pl.lit(1)).agg(a2=pl.col.a.slice(1, 1).first()).drop("literal")
+
+    q = pl.concat([q1, q2], how="horizontal").collect()
+
+    result = q.select(
+        eq=pl.col.a1.eq(pl.col.a2),
+        eq_missing=pl.col.a1.eq_missing(pl.col.a2),
+        ne=pl.col.a1.ne(pl.col.a2),
+        ne_missing=pl.col.a1.ne_missing(pl.col.a2),
+    )
+
+    assert_frame_equal(
+        result,
+        pl.select(
+            eq=pl.lit(None, pl.Boolean),
+            eq_missing=True,
+            ne=pl.lit(None, pl.Boolean),
+            ne_missing=False,
+        ),
+    )

